src/ray/gcs/tables.h

```diff
@@ -175,9 +175,23 @@ class Log : virtual public PubsubInterface<ID> {
   Status CancelNotifications(const JobID &job_id, const ID &id,
                              const ClientID &client_id);
 
+  void AddShards(std::vector<std::shared_ptr<RedisContext>> &contexts) {
```

Comment:
I'd like to get rid of this method. However, it may be easier to tackle it in a separate PR.

That would mean that we need to pass in the `contexts` into the constructor of this class. Which means we would need to connect to the primary Redis shard (to get the number of shards) before calling the table/log constructor, which means we would need to connect to the primary redis shard in the gcs client constructor, which is probably reasonable, but may cause other issues.

@stephanie-wang any thoughts about this?

Reply:
I guess this is one of the most important point causing some suspicious failures.